bfvmm/include/interface/arch/intel_x64/cpuid.h

```diff
@@ -0,0 +1,108 @@
+//
+// Copyright (C) 2019 Assured Information Security, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef BFVMM_INTERFACE_CPUID_INTEL_X64_H
+#define BFVMM_INTERFACE_CPUID_INTEL_X64_H
+
+namespace bfvmm::intel_x64
+{
+
+/// vCPU CPUID Interface
+///
+/// vCPUs that support CPUID virtualization must implement this interface
+///
+class vcpu_cpuid_interface
+{
+
+public:
+
+    /// CPUID leaf type
+    ///
+    ///
+    using leaf_t = uint64_t;
+
+    /// CPUID Add Handler
+    ///
+    /// Add a handler for a CPUID VM exit caused by the given CPUID leaf.
+    ///
+    /// @param leaf the CPUID leaf register handler @param h for
+    /// @param h the handler to call upon a VM exit
+    ///
+    VIRTUAL void cpuid_add_handler(leaf_t leaf, const handler_delegate_t &h) PURE;
```

Comment:
I changed the naming convention slightly from what existed previously with the following thoughts:

* the first part of the function identifies the interface that it belongs to: ```cpuid_```
* the rest of the function describes the use case it provides ```add_handler```

I was thinking that code completion tools would be better able to suggest available functionality if we named functions like this. For example, if you started typing vcpu->cpui..., you would see all the functions that belong to the cpuid interface.

Thoughts?

Reply:
also not an issue, see below